### PR TITLE
KAFKA-5603: Don't abort TX for zombie tasks

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
PR for 0.11 (PR for trunk is #3719)
\cc @enothereska @bbejeck @guozhangwang @dguy

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 1]
```diff
@@ -752,18 +752,29 @@ public void shouldAbortTransactionOnDirtyClosedIfEosEnabled() throws Exception {
...
             eosConfig, streamsMetrics, stateDirectory, null, time, producer);
 
-        task.close(false);
+        task.close(false, false);
         task = null;
         assertTrue(producer.transactionAborted());
     }
 
     @Test
+    public void shouldAbortTransactionOnZombieClosedIfEosEnabled() throws Exception {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: `shouldNot...` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 2]
```diff
@@ -411,14 +413,18 @@ void closeSuspended(boolean clean, RuntimeException firstException) {
...
+                            producer.abortTransaction();
+                        }
                         transactionInFlight = false;
                     } catch (final ProducerFencedException e) {
                         // can be ignored: transaction got already aborted by brokers/transactional-coordinator if this happens
                     }
                 }
                 try {
-                    recordCollector.close();
+                    if (!isZombie) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:426]: Why do not close the record collector if it is a zombie?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -411,14 +413,18 @@ void closeSuspended(boolean clean, RuntimeException firstException) {
             if (eosEnabled) {
                 if (!clean) {
                     try {
-                        producer.abortTransaction();
+                        if (!isZombie) {
+                            producer.abortTransaction();
+                        }
                         transactionInFlight = false;
                     } catch (final ProducerFencedException e) {
                         // can be ignored: transaction got already aborted by brokers/transactional-coordinator if this happens
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: since we can safely ignore the `ProducerFencedException` do we want to add a `finally` block to set `transactionInFlight` to false?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedTasksTest.java [hunk 4]
```diff
@@ -192,7 +192,7 @@ public void shouldCloseTaskOnSuspendWhenRuntimeException() {
         mockInitializedTask();
         t1.suspend();
         EasyMock.expectLastCall().andThrow(new RuntimeException("KABOOM!"));
-        t1.close(false);
+        t1.close(false, false);
         EasyMock.expectLastCall();
         EasyMock.replay(t1);
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/AssignedTasksTest.java:198]: super nit: not part of this PR proper, but maybe set `RuntimeException` returned from the `suspendTask()` call to variable then use it in the `assert` statement? Makes things a little easier to understand.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:426]: Because record collector calls `Producer.close()` and this will result in an exception -- after a  producer got fenced, you should not call any method anymore.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: It's doesn't matter as we are already closing the task here and will not read the value of `transactionInFlight` anymore.

**Comment by mjsax** [comment]
Updated this.

**Review by bbejeck — APPROVED** [review]
Thanks @mjsax, LGTM

**Comment by dguy** [comment]
Just kicking off a system test build

**Comment by dguy** [comment]
System tests pass: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-08-24--001.1503567035--mjsax--kafka-5603-dont-abort-tx-for-zombie-tasks-01101--bdff23f/report.html
LGTM

**Comment by dguy** [comment]
merged to 0.11
